Allow transactions that span epochs to commit if primary sequencer does not change.

WenbinZhu opened the pull request:
## Overview

- Currently we abort any transaction that spans epochs. But this is not necessary. If the primary sequencer remains the same (i.e. it has consecutive epochs), we could allow the transaction to commit as the sequencer has all the information.

- This could improve performance in case of flakey nodes.

- A unit test that covers two cases: primary sequencer has / does have consecutive epochs.

- Removed sequencer thread configurations and all synchronization
    modifiers in SequencerServer since it is single thread by design.

Related issue(s) (if applicable): #1767 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -434,6 +445,11 @@ public synchronized void resetServer(CorfuPayloadMsg<SequencerTailsRecoveryMsg>
             streamTailToGlobalTailMap.putAll(streamTails);
         }
 
+        // Update firstConsecutiveEpoch if the bootstrap epoch is consecutive.
```
This optimization is very specific, why don't you address the general case (i.e. track a range of epochs as opposed to just the last epoch change) ?

Every time the sequencers epoch changes, you can maintain a consecutive range [a, b] and then any epoch within [a, b] can be accepted.

WenbinZhu replied:
@Maithem Sorry for the confusion, this comment is wrong,  missed a "not", should be "if the bootstrap epoch is not consecutive". I've made the correction. 

This patch does track a range of epochs: the `firstConsecutiveEpoch` is `a` in your example, and the current sequencer epoch is `b`.  Any time the sequencer epoch changes, we update `firstConsecutiveEpoch` if the new epoch is not consecutive. If it's consecutive, we don't update `firstConsecutiveEpoch`. So in this way, the range [firstConsecutiveEpoch, sequencer epoch] is a consecutive range. Any transaction falls in this range can be accepted for further resolution.

vjeko commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -248,10 +257,12 @@ private TxResolutionResponse txnCanCommit(TxResolutionInfo txInfo) {
+        // timestamp less than the firstConsecutiveEpoch since we are sure
+        // this sequencer is always the primary sequencer after this epoch.
+        long tokenEpoch = txSnapshotTimestamp.getEpoch();
+        if (tokenEpoch != sequencerEpoch && tokenEpoch < firstConsecutiveEpoch) {
```
In what case will you end up tokenEpoch > firstConsecutiveEpoch? From what I understand, you will end up with either:

1) tokenEpoch < firstConsecutiveEpoch
2) tokenEpoch == firstConsecutiveEpoch

WenbinZhu replied:
Every time system epoch changes, the primary sequencer will get a bootstrap request. In the bootstrap handling logic, firstConsecutiveEpoch will be updated if the bootstrap epoch is a consecutive epoch to the current sequencer epoch:https://github.com/CorfuDB/CorfuDB/blob/17f8d5314d0b415a9127f48333d1605580d3b79e/infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java#L449

In this way, we keep a consecutive range: [firstConsecutiveEpoch, sequencerEpoch]. This sequencer is the primary sequencer for every epoch within this range. So when a transaction has a tokenEpoch within this range, it could be accepted. In this case, from the time when the transaction starts to the time when transaction wants to commit, the primary sequencer never changes (this is the case when we end up tokenEpoch >= firstConsecutiveEpoch).

vjeko replied:
I see. I though you were updating the `firstConsecutiveEpoch` when `bootstrapMsgEpoch == sequencerEpoch + 1`. I would be nice if you expressed this logic in terms of a sliding window/range:

`strictlyMonotonicEpochWindow = [firstConsecutiveEpoch, sequencerEpoch]`

.. and then do the boundary check in form of 'range.contains(tokenEpoch)'. You might want to take a look at: https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Range.html

WenbinZhu replied:
Actually my first thought was also using Range, but then I found the approach I'm currently using is more efficient since it only keeps a `long` variable and does a simple long comparison, instead of using `contains`, which is more costly.

vjeko replied:
You really should not worry about performance in this case, since epoch changes happen very infrequently (relative to other events in the system). I would argue that this is a premature optimization.

Readability and understandability are way more important.

WenbinZhu replied:
Yes I agree, will do.

xnull replied:
Also, you could extract `[firstConsecutiveEpoch, sequencerEpoch]` into a class (it could be ConsequtiveEpoch?) also move all related logic to that class and cover the class by unit tests.

WenbinZhu replied:
@vkeko Add a helper method for range check as discussed.

zalokhan commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -248,10 +257,12 @@ private TxResolutionResponse txnCanCommit(TxResolutionInfo txInfo) {
+        // timestamp less than the firstConsecutiveEpoch since we are sure
+        // this sequencer is always the primary sequencer after this epoch.
+        long tokenEpoch = txSnapshotTimestamp.getEpoch();
+        if (tokenEpoch != sequencerEpoch && tokenEpoch < firstConsecutiveEpoch) {
```
Shouldn't this be ```if (tokenEpoch != sequencerEpoch || tokenEpoch < firstConsecutiveEpoch)```?

WenbinZhu commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -148,11 +150,18 @@
+     * epochs from this epoch to {@link this#sequencerEpoch}
+     */
+    @Getter
+    private volatile long firstConsecutiveEpoch = Layout.INVALID_EPOCH;
```
Done.

WenbinZhu commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -248,10 +257,12 @@ private TxResolutionResponse txnCanCommit(TxResolutionInfo txInfo) {
-        if (txSnapshotTimestamp.getEpoch() != sequencerEpoch) {
-            log.debug("ABORT[{}] snapshot-ts[{}] current epoch[{}]", txInfo,
-                    txSnapshotTimestamp, sequencerEpoch);
+        // timestamp less than the firstConsecutiveEpoch since we are sure
```
Done.

WenbinZhu commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -85,6 +84,9 @@
 public class SequencerServer extends AbstractServer {
 
+    // Default cache size for transaction conflict resolution
+    private static final long DEFAULT_CACHE_SIZE = 250_000;
```
Done. Actually this default value is already exposed in CorfuServer configurations. Removed this variable since it's redundant. Moved get config logic to ServerContext.

xnull commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -148,11 +147,19 @@
+     * {@link this#sequencerEpoch}
+     */
+    @Getter
+    private volatile long epochRangeLowerBound = Layout.INVALID_EPOCH;
```
Please use AtomicLong instead of volatile

xnull replied:
And also why you need volatile if the sequencer is single threaded

WenbinZhu replied:
Will do.

xnull commented on infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java:
```diff
@@ -468,6 +468,15 @@ public long getSequencerEpoch() {
+     * @return Sequencer cache size
+     */
+    public long getSequencerCacheSize() {
+        return Long.parseLong(getServerConfig(String.class, "--sequencer-cache-size"));
```
It colud be an NPE here, please add more checks

WenbinZhu replied:
There won't be NPE here. See https://github.com/CorfuDB/CorfuDB/blob/master/infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java#L134 
We provide a default value if user not specified.

xnull replied:
What if tomorrow someone deletes that default value?
Your code must work in any case, it must be independent of some other ode somewhere, don't rely heavily on only current code, think about changes

WenbinZhu commented:
Updates:

Removed sequencer thread configurations and all synchronization
modifiers in SequencerServer (One unit test failing because the TestServerRouter is not using the server executors, which is not correct. Addressed in #1781, Once that fix is in, this PR can pass)

Please review again. Thanks.

Maithem commented:
Can you please rebase this? I think its time to check it in.

WenbinZhu commented:
@Maithem  rebased.

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -128,23 +140,18 @@ public SequencerServer(ServerContext serverContext) {
         this.serverContext = serverContext;
         Map<String, Object> opts = serverContext.getServerConfig();
         this.executor = Executors.newFixedThreadPool(
```
You can make this a single threaded executor and get rid of the variable (and merge the comments).

WenbinZhu replied:
Done.

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:
```diff
@@ -455,8 +483,10 @@ private Timer getTimer(byte reqType) {
 
-        Token token = new Token(sequencerEpoch, globalLogTail.getAndAdd(req.getNumTokens()));
-        r.sendResponse(ctx, msg, CorfuMsgType.TOKEN_RES.payloadMsg(new TokenResponse(token, Collections.emptyMap())));
+        Token token = new Token(sequencerEpoch, globalLogTail);
```
Maybe add a comment about how the tail points to an open slot and not the last written slot and therefore you return the current position and then update.

WenbinZhu replied:
Done.
